### PR TITLE
[receiver/kafka] Add raw unmarshaler

### PR DESCRIPTION
**Pull request by atoulme**
**Description:** 
Adds a raw unmarshaler for logs to consume Kafka events.

**Link to tracking Issue:**
#13252

**Testing:**
Unit tests

**Documentation:**
Added to README and testdata config.

### CODE REVIEW

**Comment by djaglowski** [comment]
cc codeowners @pavolloffay @MovieStoreGuy

File: receiver/kafkareceiver/direct_unmarshaler.go
Comment (MovieStoreGuy) [inline receiver/kafkareceiver/direct_unmarshaler.go:22]: This naming doesn't represent encoding type as in `direct logs` has no meaning to me.

Considering this is just forcing the logs as a raw message into message body, why not have this as `RawUnmarshaler`?

Comment (MovieStoreGuy) [reply receiver/kafkareceiver/direct_unmarshaler.go:22]: There should also be a Marshaller for this as well.

**Review by MovieStoreGuy — COMMENTED** [review]
Not sure I understand the problem trying to solve here

File: receiver/kafkareceiver/direct_unmarshaler.go
Comment (atoulme) [reply receiver/kafkareceiver/direct_unmarshaler.go:22]: `direct` is used by Apache Camel - I can use `raw` instead.

**Comment by atoulme** [comment]
@MovieStoreGuy regarding your comment - I am seeing Kafka (and other message queue systems) used for ingesting logs.
Here is a list of Apache Kafka use cases: https://kafka.apache.org/uses
> Many people use Kafka as a replacement for a log aggregation solution. Log aggregation typically collects physical log files off servers and puts them in a central place (a file server or HDFS perhaps) for processing. Kafka abstracts away the details of files and gives a cleaner abstraction of log or event data as a stream of messages. This allows for lower-latency processing and easier support for multiple data sources and distributed data consumption. In comparison to log-centric systems like Scribe or Flume, Kafka offers equally good performance, stronger durability guarantees due to replication, and much lower end-to-end latency.

As such, being able to ingest Kafka messages, and render them as log records, is useful for OpenTelemetry.

**Comment by MovieStoreGuy** [comment]
@atoulme awesome, this makes a whole lot more sense to me and I greatly appreciate it.

File: exporter/kafkaexporter/README.md
Comment (MovieStoreGuy) [inline exporter/kafkaexporter/README.md:26]: I feel like we should move the encoding out into a seperate doc at this point.

I also think it is worthwhile mentioned that we can't infer any of the metadata like attributes or resources using raw.

File: exporter/kafkaexporter/raw_marshaler.go
Comment (MovieStoreGuy) [inline exporter/kafkaexporter/raw_marshaler.go:21]: While I prefer this over the default encoding package, there is subtle differences between them

If we are going to use this, we have to use throughout the exporter.

File: exporter/kafkaexporter/raw_marshaler.go
Comment (MovieStoreGuy) [inline exporter/kafkaexporter/raw_marshaler.go:34]: This does impact performance, is sorted keys important here?

File: exporter/kafkaexporter/raw_marshaler.go
Comment (MovieStoreGuy) [inline exporter/kafkaexporter/raw_marshaler.go:56]: Please double check me here but it is possible for an empty message to be added to the messages right?

**Review by MovieStoreGuy — COMMENTED** [review]
Just some further comments but I agree this is a worth while change to help people adopt logs with Open Telemety

File: exporter/kafkaexporter/raw_marshaler.go
Comment (atoulme) [reply exporter/kafkaexporter/raw_marshaler.go:21]: It is used elsewhere, such as for HEC exporter. Happy to scrub it out if it helps land the PR, and maybe later open it back up with a performance benchmark.

File: exporter/kafkaexporter/raw_marshaler.go
Comment (atoulme) [reply exporter/kafkaexporter/raw_marshaler.go:56]: Yes, I think so.

File: exporter/kafkaexporter/raw_marshaler.go
Comment (atoulme) [reply exporter/kafkaexporter/raw_marshaler.go:34]: Mostly to make my tests pass reliably. If we do move away from jsoniter, that might solve itself:
https://stackoverflow.com/questions/18668652/how-to-produce-json-with-sorted-keys-in-go
> The json package always orders keys when marshalling.

File: exporter/kafkaexporter/README.md
Comment (atoulme) [reply exporter/kafkaexporter/README.md:26]: not sure where to move things. I'll add a sentence regarding metadata.

Comment (atoulme) [reply exporter/kafkaexporter/README.md:26]: This might come after this - I might ask that we move encoders to a lib and reuse them across multiple receivers/exporters. Useful for my work on event hubs for Azure.

File: exporter/kafkaexporter/raw_marshaler.go
Comment (MovieStoreGuy) [inline exporter/kafkaexporter/raw_marshaler.go:50]: I suggest that we avoid sending an empty buffer to avoid that overhead on kafka's side.

**Review by MovieStoreGuy — APPROVED** [review]
I am happy with these changes, just a minor nit on ensuring we don't send empty buffers but otherwise, I am okay with the changes.

File: exporter/kafkaexporter/raw_marshaler.go
Comment (atoulme) [reply exporter/kafkaexporter/raw_marshaler.go]: OK fixed

**Review by MovieStoreGuy — APPROVED** [review]

**Comment by MovieStoreGuy** [comment]
@bogdandrutu , I have a approved it as a code owner, but could I ask you to review it as a maintainer?

**Review by jpkrohling — APPROVED** [review]
Sorry for the delay in reviewing this. I won't merge right now as we have a release in progress, so, this might need another rebase soon, but LGTM.

**Review by pavolloffay — APPROVED** [review]

**Pull request merged**
